### ISSUE TITLE
Best way to save and load measurements from external data store

### ISSUE THREAD

**Issue by timoangerer** [OHIF/Viewers#1800]
Hi,

I'm trying to build an annotation tool to draw rectangular ROIs on images and allow for saving/loading of the measurements. The measurements should **not** be saved inside DICOM SR, but in an external database instead.

### What is the best way to:

1. Save measurements in an external data store (e.g MongoDB).
2. Load measurements back into the viewer and correctly display the measurements inside the viewer and the SidePanel.

Which functions, services or redux actions should I look into to implement this in the best way possible?

#### What I tried/looked into:

For point one, I've got 2 vague ideas, but I'm not sure if they make sense:

- Export content of the redux state at `State > timepointManager > measurements`.
- Export content of `cornerstoneTools.globalImageIdSpecificToolStateManager.toolState`.

For point two, I looked through the redux store using the redux-dev tools and found the `SET_MEASUREMENTS` action. When I dispatch this action with a measurement, the store is updated correctly, but the measurement is not rendered inside the viewer.

#### Similar issues:

In #1102, @dannyrb suggestes to "hydrate" cornerstone's toolstate.
In #1451, @dannyrb suggestes to "hook directy into the tool state".

Any help is greatly appreciated!

**Comment by pieper** [comment]
Are you avoiding SR because you think it's overly complex or something? (If you think that, I have to agree).  But it is an international standard and if your goal is to just save and restore in json formatted measurements in a noSQL database you can use the [dicomweb-server](https://github.com/dcmjs-org/dicomweb-server) that already stores in CouchDB in json.

Just to day @JamesAPetts and @swederik [merged code](https://github.com/dcmjs-org/dcmjs/commit/0e1e2921e953bde8f4707b025f584a6c558ffedd) for length, arrow, bidirectional, and ellipse tools, so adding rectangle would just follow the same pattern.

**Comment by JamesAPetts** [comment]
I second that adding a rectangle is not at all hard, I simply only implemented what my client wanted. I'd be happy to review a PR for IO of Rectangle to/from `cornerstoneTools` for `dcmjs`, it should be straightforward. If you make one, tag me :).

**Comment by timoangerer** [comment]
Thank you very much for the responses!

SR does indeed look quite intimidating and complicated for my use case. I just want to allow different students to annotate the same image and save those annotations in a DB. I'm not sure if it is possible to store something like an "annotator name" inside SR. Then I would like to restore all the annotations inside the viewer so an expert can double-check them.

After thinking about it, I'd rather do the annotation with a single point instead of a rectangle.

Regarding the dicomweb-server suggestion: I have to use Orthanc as backend for this project.

---

Do you think I should go with SR given my simple use case?

And how could I do the saving and restoring without using SR?

**Comment by pieper** [comment]
@timoangerer - maybe for what you are describing you could make use of the [Crowds Cure](https://www.crowds-cure.org/) system that we ran at RSNA in 2017 and 2018.  It's pretty close to what you describe.  The site is still live so you can create an anonymous account and try it out.

The architecture includes user logins, a noSQL database, a DICOMweb server, and annotations saved in json (you can look at the result files on TCIA).  All the [code is on github](https://github.com/crowds-cure).  It even works pretty well on phones and tablets.

It doesn't currently include an easy way to review the annotations within the system; we currently use other tools to do that, for example to batch convert the JSON to proper SR (yes, annotator information and other important information can be saved in SR).

**Comment by timoangerer** [comment]
@pieper - This project does indeed look similar to what I'm trying to accomplish. Thank you for sharing it!

[This function](https://github.com/crowds-cure/cancer/blob/bf7f574d45f5f2078835550a6cb7c69a60f7bd4d/src/viewer/lib/saveMeasurementToDatabase.js#L97) seems to do the saving of the measurements. But where does the measurements object that is passed to the function come from?

I'm currently trying to understand how the `cornerstoneTools` object works. I'm assuming that I could save the contents of `cornerstoneTools.globalImageIdSpecificToolStateManager.toolState` and later load those measurements back in using the `cornerstoneTools.globalImageIdSpecificToolStateManager.restoreToolState()` function. Unfortunatily it does not seem to work this way.

What am I missing? Does it even make sense to do it this way?

**Comment by timoangerer** [comment]
I just realized that the method from my previous answer does indeed work: It adds the measurements to the viewer.

However, it does not add the measurements to the measurements SidePanel.

I guess I have to do this save/load operations a bit "higher up" in the architecture.

**Comment by pieper** [comment]
I'm not up on the details of that part.  Hope someone else can answer.

**Comment by yash0906** [comment]
@timoangerer 
i am also trying the do the same thing. 
did you add mapping for the rectangleRoi in extention/cornerstone/src/init.js 
They are only supporting length now. If you have made any progress, please share.
TIA^_^

**Comment by wangning08115** [comment]
@timoangerer    Have you solved the problem?
I have the same problem.
Can you help me?

**Comment by timoangerer** [comment]
@yash0906 & @wangning08115 I'll let you know as soon as I found a solution

**Comment by yash0906** [comment]
@timoangerer thanks for your support 
https://github.com/OHIF/Viewers/issues/1826
do you have any idea regarding this issue

**Comment by yesid08** [comment]
I think that maybe I can conttribute a little bit. If you want to obtain all the annotations of the current study you can do the following:

OHIF.measurements.MeasurementApi.Instance.tools

That will return an object with all the tools and each tool is an array with the annotatios made by the user.

Now, if you already have something like an array or annotations and you want to update the viewer with such annotations, then you should use something like this:
const localMeasurementAPI = OHIF.measurements.MeasurementApi.Instance;
// Apply Measurements ------------------ //
  tempMeasurements.forEach(tempMeasurements =>
    localMeasurementAPI.addMeasurement(
      tempMeasurements.toolType,
      tempMeasurements
    )
  );
  // Sync Measurements -------------------//
  localMeasurementAPI.syncMeasurementsAndToolData();
  cornerstone.getEnabledElements().forEach(enabledElement => {
    cornerstone.updateImage(enabledElement.element);
  });
  // Let others know that the measurements are updated
  localMeasurementAPI.onMeasurementsUpdated();
};

**Comment by yesid08** [comment]
Do you guys now which event or listener is triggered when the Viewer has already loaded an image on the screen? I want to trigger a little script when the app loads an image on the screen, but I don't really know which Cornerstone element will be the responsible for this part and also which event or listener is the correct one for this case, any ideas?
 I've already tried with the pre-registration method from extensions but the problem I have is that I need that cornerstone and OHIF loads completely...

I hope that you can help me, greetings!

**Comment by yash0906** [comment]
> I think that maybe I can conttribute a little bit. If you want to obtain all the annotations of the current study you can do the following:
> 
> OHIF.measurements.MeasurementApi.Instance.tools
> 
> That will return an object with all the tools and each tool is an array with the annotatios made by the user.
> 
> Now, if you already have something like an array or annotations and you want to update the viewer with such annotations, then you should use something like this:
> const localMeasurementAPI = OHIF.measurements.MeasurementApi.Instance;
> // Apply Measurements ------------------ //
> tempMeasurements.forEach(tempMeasurements =>
> localMeasurementAPI.addMeasurement(
> tempMeasurements.toolType,
> tempMeasurements
> )
> );
> // Sync Measurements -------------------//
> localMeasurementAPI.syncMeasurementsAndToolData();
> cornerstone.getEnabledElements().forEach(enabledElement => {
> cornerstone.updateImage(enabledElement.element);
> });
> // Let others know that the measurements are updated
> localMeasurementAPI.onMeasurementsUpdated();
> };

this thing only work for a single frame image.
If you have annotations saved on multiple frames, this will load all the annotations on a single frame only.

**Comment by smilegok** [comment]
//with anywhere you want to store data on local storage if you wish to store at DB then do it here
const toolState = cornerstoneTools.globalImageIdSpecificToolStateManager.saveToolState();
localStorage.setItem('toolState', JSON.stringify(toolState));


//then load it bak with code below in MeasurementApi.js
retrieveMeasurements(PatientID, timepointIds) {
    const retrievalFn = configuration.dataExchange.retrieve;
    const { server } = configuration;
    if (typeof retrievalFn !== 'function') {
      log.error('Measurement retrieval function has not been configured.');
      return;
    }

    return new Promise((resolve, reject) => {
      retrievalFn(server).then(measurementData => {
        if (measurementData) {
          log.info('Measurement data retrieval');
          log.info(measurementData);
          debugger;

          let localtoolState = localStorage.getItem('toolState');
          if(!localtoolState)
          {
            return;
          }
          let toolState = JSON.parse(localtoolState)
          cornerstoneTools.globalImageIdSpecificToolStateManager.restoreToolState(toolState);

          Object.keys(toolState).forEach(KesitId => {
            const kesits = toolState[KesitId];
            Object.keys(kesits).forEach(measurementTypeId => {
              const measurements = kesits[measurementTypeId]["data"];
              Object.keys(measurements).forEach(DataId => {
                const measurement =  measurements[DataId]
                this.addMeasurement(measurementTypeId, measurement);
              });
            });
          });
        }

        resolve();

        // Synchronize the new tool data
        this.syncMeasurementsAndToolData();

        cornerstone.getEnabledElements().forEach(enabledElement => {
          if (enabledElement.image) {
            cornerstone.updateImage(enabledElement.element);
          }
        });

        // Let others know that the measurements are updated
        this.onMeasurementsUpdated();
      }, reject);
    });
  }

**Comment by dpostolovski** [comment]
Perhaps a better, modular way would be to change the configuration of the measurements API and implement the retrieve and store functions:

```
OHIF.measurements.MeasurementApi.setConfiguration({
      dataExchange: {
        retrieve: DICOMSR.retrieveMeasurements, // implement a function with this signature and place it here
        store: DICOMSR.storeMeasurements,
      },
      server,
    });
```

**Comment by sedghi** [comment]
You can look at measurementService on how you can hydrate the annotations that you retrieve

platform/core/src/services/MeasurementService/MeasurementService.ts

**Comment by phoenixsac** [comment]
@sedghi can you give some pointers on how to hydrate it? I saved the measurement data to the backend, then loaded the stored measurement json data from my backend and called the addRawMeasurements [method](https://github.com/OHIF/Viewers/blob/47846387177b4502936d7c506508a3b7f76c66bc/platform/core/src/services/MeasurementService/MeasurementService.ts#L103), but it throws errors like source not defined and no measurement mappings found. Any suggestions would be really helpful.

**Comment by amaguado** [comment]
hello, is there any update on this matter?

**Comment by sedghi** [comment]
see how cornerstone3D hydrate example https://www.cornerstonejs.org/live-examples/dynamicallyaddannotations

**Comment by eric1lubow** [comment]
I've resolved it. 
![intro](https://github.com/user-attachments/assets/565e5607-63ab-4ee2-8141-4e851c512f7b)

**Comment by nicolay-r** [comment]
@timoangerer, @phoenixsac, @amaguado @sedghi 

I came across this topic with the need for importing measurements that been previously added and exported.

By spending quite a while, I've attempted to reprodue importing using suggested `measurementService.addRawMeasurement`, with implementation idea borrowed from `extensions/cornerstone-dicom-sr/src/utils/hydrateStructuredReport.ts`.

My takeaway on the most challenge was to format `annotations` such that it meet requirements for imported measurements / vidualizaiton via cornerstone (cached data).

I've tested on `Bidirectional` tool, while the others might require extra parameters in `data` and `handles`.

**gist here:** https://gist.github.com/nicolay-r/acaa75c183ee47868c3feddbf92433f7
**full log**: https://nicolayr.com/blog/articles/2026-03/ohif-import-measurements

![](https://github.com/user-attachments/assets/3f522c1b-7bd1-4423-a5c0-8995760cc9d3)
